README
This distribution contains the module Filesys::DfPortable

Filesys::DfPortable - Perl extension for portable filesystem information.

SYNOPSIS


  use Filesys::DfPortable;

  my $ref = dfportable("C:\\"); # Default block size is 1, which outputs bytes
  if(defined($ref)) {
     print"Total bytes: $ref->{blocks}\n";
     print"Total bytes free: $ref->{bfree}\n";
     print"Total bytes avail to me: $ref->{bavail}\n";
     print"Total bytes used: $ref->{bused}\n";
     print"Percent full: $ref->{per}\n"
  }
                                                                                                                
  my $ref = dfportable("/tmp", 1024); # Display output in 1K blocks
  if(defined($ref)) {
     print"Total 1k blocks: $ref->{blocks}\n";
     print"Total 1k blocks free: $ref->{bfree}\n";
     print"Total 1k blocks avail to me: $ref->{bavail}\n";
     print"Total 1k blocks used: $ref->{bused}\n";
     print"Percent full: $ref->{per}\n"
  }


DESCRIPTION

This module provides a portable way to obtain filesystem information.

The module should work with all versions of Windows (95 and up),
all flavors of Unix, including Mac OS X (Darwin, Tiger, etc), and Cygwin.

dfportable() requires a directory argument that represents the filesystem
you want to query. There is also an optional block size argument so the
you can tailor the size of the values returned. The default for block
size is 1, this will cause the function to return the values in bytes.
If you never use the block size argument, then you can think of any
instance of "blocks" in this document to really mean "bytes".

dfportable() returns a reference to a hash. The keys available in
the hash are as follows:

{blocks} = Total blocks on the filesystem.

{bfree} = Total blocks free on the filesystem.

{bavail} = Total blocks available to the user executing the Perl
application. This can be different than bfree if you have per-user
quotas on the filesystem, or if the super user has a reserved amount.
bavail can also be a negative value because of this. For instance
if there is more space being used then you have available to you.

{bused} = Total blocks used on the filesystem.

{per} = Percent of disk space used. This is based on the disk space
available to the user executing the application. In other words, if
the filesystem has 10% of its space reserved for the superuser, then
the percent used can go up to 110%.


You can obtain inode information through the module as well. But you
must call exists() on the {files} key to make sure the information is
available:
if(exists($ref->{files})) {
	#### Inode info is available
}
Some filesystems may not return inode information, for
example Windows, and some NFS filesystems.

Here are the available inode keys:

{files} = Total inodes on the filesystem.

{ffree} = Total inodes free on the filesystem.

{favail} = Total inodes available to the user executing the application.
See the rules for the {bavail} key.

{fused} = Total inodes used on the filesystem.

{fper} = Percent of inodes used on the filesystem. See rules for the {per}
key.


If the dfportable() call fails for any reason, it will return
undef. This will probably happen if you do anything crazy like try
to get information for /proc, or if you pass an invalid filesystem name,
or if there is an internal error. dfportable() will croak() if you pass
it a undefined value.

Requirements:
Your system must contain either statvfs(), statfs(), GetDiskFreeSpaceA(), 
or GetDiskFreeSpaceEx().

You must be running Perl 5.6 or higher.


INSTALL
TO INSTALL RUN:
	
	perl Makefile.PL
	make
	make test
	make install


(On Windows you will probably be using nmake or a ppm file)

During the 'make test', test.pl will try to test with '/' if you are
using a Unix system or 'C:\' if you are using Windows. 

Once installed, run 'perldoc Filesys::DfPortable' for more information. 

If you have any problems or questions please email me at IGuthrie@aol.com
with "DfPortable" in the subject line. If you run into a build problem, 
please include the output of the install commands, the version of Perl
you are using (perl -v), and what operating system you are using.

Copyright (c) 2006 Ian Guthrie. All rights reserved.
               This program is free software; you can redistribute it and/or
               modify it under the same terms as Perl itself.


Changes
Revision history for Perl extension Filesys::DfPortable
                                                                                                                             
0.81  Wed May 10 22:56:11 2006
        - Module creation

0.82  Wed May 10 23:10:11 2006
        - Oops had statfs header info backward in Makefile.PL

0.83  Thu May 11 17:00:11 2006
        - Changed interface to return a reference to hash instead of a list
	- Added hash values percent and used
	- Stopped setting $!

0.84  Sat May 13 21:55:11 2006
        - Added inode information
	- Fixed a possible bug with user blocks


test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..2\n"; }
END {print "not ok 1\n" unless $loaded;}
require 5.006;
use Config qw(%Config);
use Filesys::DfPortable;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

my $dir;
($Config{osname} =~ /^MSWin/i)  and
	$dir = "C:\\" or
	$dir = "/";

my $ref = Filesys::DfPortable::dfportable($dir);

defined($ref) and
	print"ok 2\n" or
	die "not ok 2\nstatvfs\(\) call failed for \"$dir\" $!\n";

print"Results for directory: \"$dir\" in bytes:\n";
print "Total: $ref->{blocks}\n";
print "Free: $ref->{bfree}\n";
print "Available: $ref->{bavail}\n";
print "Used: $ref->{bused}\n";
print "Percent Full: $ref->{per}\n";


print"All tests successful!\n\n";



Makefile.PL
use ExtUtils::MakeMaker;
use Config qw(%Config);

#### Build information
my $statvfs_header = "sys/statvfs.h";
my $statfs_header = "sys/statfs.h";
my $statvfs_symbol = "d_statvfs";
my $statvfs_header_def = "i_sysstatvfs";
my $statfs_symbol = "d_statfs_s";  #### Really the struct def, we equate it to statfs being defined
my $statfs_header_def = "i_sysstatfs";
my $define = "";
my $statfs_use_mount = 0;
my $xs_file = 'DfPortable.xs';


print "OS = $Config{osname}\n";

#### Windows
if($Config{osname} =~ /^MSWin/i) {
	####$define .= "-DDFPORTABLE_WIN ";
	copy_xs("XS_win", $xs_file);
	print "Building for Windows ....\n";
}

#### Check for the existance of statvfs
elsif(check_statvfs()) {
	####$define .= "-DDFPORTABLE_STATVFS ";
	copy_xs("XS_statvfs", $xs_file);
	print "Building with statvfs ....\n";
}

#### Check for the existance of statfs
elsif(check_statfs()) {
	#### use_mount needed for headers
	($statfs_use_mount) &&
		($define .= "-DDFPORTABLE_STATFS ") ||
		($define .= "-DDFPORTABLE_STATFS_USE_MOUNT ");

	#### Needed for 4 arg statfs
	($Config{osname} =~ /^solaris$/i) &&
        ($define .= "-DDFPORTABLE_SOLARIS ");
	
	copy_xs("XS_statfs", $xs_file);
	print "Building with statfs ....\n";
}

#### OS/2, old Mac, etc
else {
	print "We could not find statvfs, statfs, or GetDiskFreeSpace.\n";
	die "You need at least one of these to build this module.\n";
}



sub check_statvfs {
	print "Checking for statvfs .....\n";
	if(exists $Config{$statvfs_symbol} && defined $Config{$statvfs_symbol}) { 
		print "$statvfs_symbol is defined.\n";
		if(exists $Config{$statvfs_header_def} && defined $Config{$statvfs_header_def}) {
			print "$statvfs_header_def is defined.\n";
			return(1);
		}

		else {
			print "Weird, $statvfs_header_def is not defined.\n";
			#### Have never seen a system with statvfs and no sys/statvfs.h header
			#### Lets see if we can find one
			if(look_for_header($statvfs_header)) {
				return(1);
			}

			else {
				#### no idea what header would be
				print "Cannot find a $statvfs_header file\n";
				print "We will not try to build with statvfs\n";
				return(0);
			}
		}
	}

	else {
		print "$statvfs_symbol is not defined\n";

		### OK if we find a header should we build with it?
		if(look_for_header($statvfs_header)) {
			return(1);
		}

		else {
			#### don't use statvfs
			print "Cannot find a $statvfs_header file\n";
			print "We will not try to build with statvfs\n";
			return(0);
		}
	}

	return(0);
}


sub check_statfs {
	print "Checking for statfs .....\n";
	if(exists $Config{$statfs_symbol} && defined $Config{$statfs_symbol}) {
		print "Good, $statfs_symbol is defined.\n";
		if(exists $Config{$statfs_header_def} && defined $Config{$statfs_header_def}) {
			print "$statfs_header_def is defined.\n";
			return(1);
		}

		else {
			print "$statfs_header_def not defined.\n";
			#### check for BSD and Darwin 
			if($Config{osname} =~ /^darwin|^bsd|bsd$/i) {
				print "You are running Darwin or BSD.\n";
				print "Will assume you need the mount.h and param.h headers.\n";
				$statfs_use_mount = 1;
				return(1);
			}

			elsif(look_for_header($statfs_header)) {
				return(1);
			}

			else {
				#### don't use statfs
				print "Cannot find a $statfs_header file\n";
				print "We will not try to build Statfs\n";
			}
		}
	}

	else {
		print "$statfs_symbol is not defined\n";

		if(look_for_header($statfs_header)) {
			return(1);
		}
	
		else {
			print "Cannot find a $statfs_header file\n";
			print "We will not try to build Statfs\n";
			return(0);
		}
	}

	return(0);
}


sub copy_xs {
my $source = shift;
my $dest = shift;

	open(SOURCE, "$source") or die "$! $source\n";
	open(DEST, ">$dest") or die "$! $dest\n";
	@contents = <SOURCE>;
	print DEST @contents;
	close(DEST);
	close(SOURCE);
}


sub look_for_header {
my $header = shift;

  	my @header_inc = split(/\s+/, join(" ", $Config{usrinc}, $Config{locincpth}));
	foreach $header_path (@header_inc) {
		if(-f $header_path . '/' . $header) {
			print "Header found:" , $header_path . '/' . $header, "\n";
			return(1);
		}
	}

	return(0);
}


WriteMakefile(
    'NAME'	=> 'Filesys::DfPortable',
    'VERSION_FROM' => 'DfPortable.pm', # finds $VERSION
    'LIBS'	=> [''],   # e.g., '-lm' 
    'DEFINE'	=> $define,     # e.g., '-DHAVE_SOMETHING' 
    'INC'	=> '',     # e.g., '-I/usr/include/other' 
    'clean'     => {FILES => 'DfPortable.xs'},
    'XSPROTOARG' => '-prototypes' 
);


DfPortable.pm
package Filesys::DfPortable;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);
use Carp;
require Exporter;
require DynaLoader;
require 5.006;

@ISA = qw(Exporter DynaLoader);
@EXPORT = qw(dfportable);
$VERSION = '0.84';
bootstrap Filesys::DfPortable $VERSION;

sub dfportable {
my ($dir, $block_size) = @_;
my ($used, $fused);
my ($per, $fper);
my ($user_blocks, $user_used);
my ($user_files, $user_fused);
my %fs = ();


	(defined($dir)) ||
		(croak "Usage: dfportable\(\$dir\) or dfportable\(\$dir\, \$block_size)");

	#### If no requested block size then we will return the values in bytes
	($block_size) ||
		($block_size = 1);

	my ($frsize, $blocks, $bfree, $bavail, $files, $ffree, $favail) = _dfportable($dir);

	#### Some system or XS failure, something like /proc, or bad $dir
	if($frsize == 0 || $blocks == 0) {
		return();
	}

	#### Change to requested or default block size
	if($block_size > $frsize) {
		my $result = $block_size / $frsize;
		$blocks /= $result;
		($bfree != 0) &&
			($bfree /= $result);
		#### Keep bavail -
		($bavail < 0) &&
			($result *= -1);

		($bavail != 0) &&
			($bavail /= $result);
	}

	elsif($block_size < $frsize) {
		my $result = $frsize / $block_size;
		$blocks *= $result;
		$bfree *= $result;
		#### Keep bavail -
		($bavail < 0) &&
			($result *= -1);
		$bavail *= $result;
	}

	$used = $blocks - $bfree;

	#### There is a reserved amount for the su
	#### or there are disk quotas
        if($bfree > $bavail) {
                $user_blocks = $blocks - ($bfree - $bavail);
                $user_used = $user_blocks - $bavail;
                if($bavail < 0) {
                        #### over 100%
                        my $tmp_bavail = $bavail;
                        $per = ($tmp_bavail *= -1) / $user_blocks;
                }
                                                                                                         
                else {
			if($user_used == 0) {
				$per = 0;
			}

			else {
                        	$per = $user_used / $user_blocks;
			}
                }
        }
                                                                                                         
        #### No reserved amount or quotas
        else {
                if($used == 0)  {
                        $per = 0;
                }
                                                                                                         
                else {
                        $per = $used / $blocks;
			$user_blocks = $blocks;
			$user_used = $used;
                }
        }

	#### round
        $per *= 100;
        $per += .5;
                                                                                                         
        #### over 100%
        ($bavail < 0) &&
                ($per += 100);

        $fs{per}     = int($per);
	$fs{blocks}  = $blocks;
	$fs{bfree}   = $bfree;
	$fs{bavail}  = $bavail;
	$fs{bused}   = $used;



	#### Handle inodes if system supports them
	if(defined $files && $files > 0) {
		$fused = $files - $ffree;
                #### There is a reserved amount
                if($ffree > $favail) {
                        $user_files = $files - ($ffree - $favail);
                        $user_fused = $user_files - $favail;
                        if($favail < 0)  {
                                #### over 100%
                                my $tmp_favail = $favail;
                                $fper = ($tmp_favail *= -1) / $user_files;
                        }
                                                                                                             
                        else {
				if($user_fused == 0) {
					$fper = 0;
				}

				else {
                                	$fper = $user_fused / $user_files;
				}
                        }
                }
                                                                                                             
                #### su and user amount are the same
                else {
                        if($fused == 0) {
                                $fper = 0;
                        }
                                                                                                             
                        else {
                                $fper = $fused / $files;
                        }
                                                                                                             
                        $user_files = $files;
                        $user_fused = $fused;
                }

                #### round
                $fper *= 100;
                $fper += .5;
                                                                                                             
                #### over 100%
                ($favail < 0) &&
                        ($fper += 100);

		$fs{fper}        = int($fper);
                $fs{files}       = $files;
                $fs{ffree}       = $ffree;
                $fs{favail}      = $favail;
                $fs{fused}       = $fused;
                #$fs{user_fused} = $user_fused;
                #$fs{user_files} = $user_files;
        }
                                                                                                             
        #### No valid inode info. Probably Windows or NFS
	#### Instead of undefing, just have the user call exists().
        #else {
        #        $fs{fper}        = undef;
        #        $fs{files}       = undef;
        #        $fs{ffree}       = undef;
        #        $fs{favail}      = undef;
        #        $fs{fused}       = undef;
        #        $fs{user_fused}  = undef;
        #        $fs{user_files}  = undef;
        #}
                                                                                                             

	return(\%fs);
}

1;
__END__

=head1 NAME

Filesys::DfPortable - Perl extension for filesystem space.

=head1 SYNOPSIS


  use Filesys::DfPortable;

  my $ref = dfportable("C:\\"); # Default block size is 1, which outputs bytes
  if(defined($ref)) {
     print"Total bytes: $ref->{blocks}\n";
     print"Total bytes free: $ref->{bfree}\n";
     print"Total bytes avail to me: $ref->{bavail}\n";
     print"Total bytes used: $ref->{bused}\n";
     print"Percent full: $ref->{per}\n"
  }
                                                                                                                
                                                                                                                
  my $ref = dfportable("/tmp", 1024); # Display output in 1K blocks
  if(defined($ref)) {
     print"Total 1k blocks: $ref->{blocks}\n";
     print"Total 1k blocks free: $ref->{bfree}\n";
     print"Total 1k blocks avail to me: $ref->{bavail}\n";
     print"Total 1k blocks used: $ref->{bused}\n";
     print"Percent full: $ref->{per}\n"
  }



=head1 DESCRIPTION

This module provides a portable way to obtain filesystem 
disk space information.

The module should work with all versions of Windows (95 and up),
all flavors of Unix, including Mac OS X (Darwin, Tiger, etc), and Cygwin.

dfportable() requires a directory argument that represents the filesystem
you want to query. There is also an optional block size argument so the
you can tailor the size of the values returned. The default for block
size is 1, this will cause the function to return the values in bytes.
If you never use the block size argument, then you can think of any
instance of "blocks" in this document to really mean "bytes". 

dfportable() returns a reference to a hash. The keys available in 
the hash are as follows:

{blocks} = Total blocks on the filesystem.

{bfree} = Total blocks free on the filesystem.

{bavail} = Total blocks available to the user executing the Perl 
application. This can be different than bfree if you have per-user 
quotas on the filesystem, or if the super user has a reserved amount.
bavail can also be a negative value because of this. For instance
if there is more space being used then you have available to you.

{bused} = Total blocks used on the filesystem.

{per} = Percent of disk space used. This is based on the disk space
available to the user executing the application. In other words, if
the filesystem has 10% of its space reserved for the superuser, then
the percent used can go up to 110%.

You can obtain inode information through the module as well. But you
must call exists() on the {files} key to make sure the information is
available:
if(exists($ref->{files})) {
        #### Inode info is available
}
Some filesystems may not return inode information, for
example Windows, and some NFS filesystems.
                                                                                                             
Here are the available inode keys:

{files} = Total inodes on the filesystem.

{ffree} = Total inodes free on the filesystem.

{favail} = Total inodes available to the user executing the application.
See the rules for the {bavail} key.

{fused} = Total inodes used on the filesystem.

{fper} = Percent of inodes used on the filesystem. See rules for the {per}
key.


If the dfportable() call fails for any reason, it will return
undef. This will probably happen if you do anything crazy like try
to get information for /proc, or if you pass an invalid filesystem name,
or if there is an internal error. dfportable() will croak() if you pass
it a undefined value.


Requirements:
Your system must contain statvfs(), statfs(), GetDiskFreeSpaceA(), or GetDiskFreeSpaceEx().
You must be running Perl 5.6 or higher.


=head1 AUTHOR

Ian Guthrie
IGuthrie@aol.com

Copyright (c) 2006 Ian Guthrie. All rights reserved.
               This program is free software; you can redistribute it and/or
               modify it under the same terms as Perl itself.

=head1 SEE ALSO

statvfs(2), df(1M), statfs(1M), GetDiskFreeSpaceA, GetDiskFreeSpaceEx

perl(1).

=cut
